### PR TITLE
Make upscaling images opt-in

### PR DESCRIPTION
**Pull request by dnfield**
## Description

We should be discouraging users from upscaling images, which causes increased memory usage. This problem arises because users typically do not know the intrinsic dimensions of the compressed image data, and may be providing cache width/height values that exceed the dimensions of the image.

Once this lands, we can default the parameter to false in the engine as well.

/cc @zanderso @chinmaygarde FYI

## Related Issues

Part of https://github.com/flutter/flutter/issues/59578
Engine side PR was https://github.com/flutter/engine/pull/19067

## Tests

Test that upscaling still works when the parameter is set to true
Test that downscaling works (previously untested)
Test that upscaling is not applied when the parameter is set to false.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.

I'll do a FROB run to make sure this breaks nothing internally.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/painting/binding.dart [hunk 1]
```diff
@@ -84,13 +84,15 @@ mixin PaintingBinding on BindingBase, ServicesBinding {
   Future<ui.Codec> instantiateImageCodec(Uint8List bytes, {
     int cacheWidth,
     int cacheHeight,
+    bool allowUpscaling = false,
```
Comment (goderbauer) [inline packages/flutter/lib/src/painting/binding.dart:87]: Can you add a sentence about this parameter to the doc comment above?

**Review by goderbauer — APPROVED** [review]
LGTM

**Comment by goderbauer** [comment]
Cirrus failure are presumably resolved when the engine roll lands?

**Comment by dnfield** [comment]
Analysis issues are places in the code where I failed to fix up signatures. Will take  a look and add the doc comment.

File: packages/flutter/lib/src/painting/binding.dart (hunk 1 above)
Comment (dnfield) [reply packages/flutter/lib/src/painting/binding.dart:95]: I updated the docstring in general here - there were some old no-longer-existant things in there and some formatting issues, as well as something that's changed about aspect ratio preservation.  PTAL

**Comment by dnfield** [comment]
My FROB run had lots of failures, but none of them seem related to this change. That's the best I can make of that :\

**Review by goderbauer — APPROVED** [review]
Still LGTM

Thanks for the updated docs!

**Pull request merged**